pkg/tools/cli/loadtest/service/loader.go

```diff
@@ -138,45 +165,101 @@ func (l *loader) Do(ctx context.Context) <-chan error {
+		log.Infof("start load test(%s, %s)", l.service.String(), l.operation.String())
 		defer close(ech)
-		eg, egctx := errgroup.New(ctx)
-		eg.Limitation(l.concurrency)
-		for _, req := range l.requests {
-			r := req
-			eg.Go(safety.RecoverFunc(
-				func() error {
-					_, err := l.client.Do(egctx, l.addr, func(ctx context.Context, conn *grpc.ClientConn, copts ...grpc.CallOption) (interface{}, error) {
-						err := l.loaderFunc(ctx, vald.NewValdClient(conn), r, copts...)
-						atomic.AddInt32(&pgCnt, 1)
-						if err != nil {
-							log.Warn("an error occurred while executing loaderfunc:", err)
-							atomic.AddInt32(&errCnt, 1)
-						}
-						return nil, err
-					})
-					if err != nil {
-						select {
-						case <-ctx.Done():
-							ech <- errors.Wrap(err, ctx.Err().Error())
-						case ech <- err:
-						}
-					}
-					return nil
-				}))
+		defer ticker.Stop()
+		start = time.Now()
+		err := l.do(ctx, f, finalize)
+		end = time.Now()
+
+		if errCnt > 0 {
+			log.Warnf("Error ratio: %.2f%%", float64(errCnt)/float64(pgCnt)*100)
 		}
-		if err := eg.Wait(); err != nil {
-			select {
-			case <-ctx.Done():
-				ech <- errors.Wrap(err, ctx.Err().Error())
-			case ech <- err:
-			}
+		if err != nil {
+			finalize(ctx, err)
 			return p.Signal(syscall.SIGKILL) // TODO: #403
 		}
-		log.Infof("Error ratio: %.2f%%", float64(errCnt)/float64(pgCnt)*100)
+		log.Infof("result:%s\t%d\t%d\t%f", l.service.String(), l.concurrency, l.batchSize, vps(int(pgCnt)*l.batchSize, start, end))
+
 		return p.Signal(syscall.SIGTERM) // TODO: #403
 	}))
 	return ech
 }
+
+func (l *loader) do(ctx context.Context, f func(res interface{}, err error), notify func(context.Context, error)) (err error) {
+	eg, egctx := errgroup.New(ctx)
+
+	switch l.operation {
+	case config.StreamInsert, config.StreamSearch:
+		var newData func() interface{}
+		switch l.operation {
+		case config.StreamInsert:
+			newData = func() interface{} {
+				return new(payload.Empty)
+			}
+		case config.StreamSearch:
+			newData = func() interface{} {
+				return new(payload.Search_Response)
+			}
+		}
+		eg.Go(safety.RecoverFunc(func() error {
+			// TODO: related to #557
+			/*
+				_, err := l.client.Do(egctx, l.addr, func(ctx context.Context, conn *grpc.ClientConn, copts ...grpc.CallOption) (interface{}, error) {
+					st, err := l.loaderFunc(ctx, conn, nil, copts...)
+					if err != nil {
+						return nil, err
+					}
+					return nil, grpc.BidirectionalStreamClient(st.(grpc.ClientStream), l.dataProvider, newData, f)
+				})
+			*/
+			conn, err := grpc.Dial(l.addr, grpc.WithInsecure())
+			if err != nil {
+				notify(egctx, err)
+				return nil
+			}
+			defer notify(egctx, conn.Close())
+			st, err := l.loaderFunc(egctx, conn, nil)
+			if err != nil {
+				notify(egctx, err)
+				return nil
+			}
+			if err := igrpc.BidirectionalStreamClient(st.(grpc.ClientStream), l.dataProvider, newData, f); err != nil {
+				notify(egctx, err)
+			}
+			return nil
+		}))
```

Comment:
sorry to send a lot of comment :bow::bow::bow:
I think you can aggregate the notify error.

```suggestion
		eg.Go(safety.RecoverFunc(func() (err error) {
			defer func() {
				if err != nil {
					notify(egctx, err)
					err = nil
				}
			}()

			conn, err := grpc.Dial(l.addr, grpc.WithInsecure())
			if err != nil {
				return err
			}
			defer notify(egctx, conn.Close())

			st, err := l.loaderFunc(egctx, conn, nil)
			if err != nil {
				return err
			}

			if err := igrpc.BidirectionalStreamClient(st.(grpc.ClientStream), l.dataProvider, newData, f); err != nil {
				return err
			}
			return nil
		}))
```